File: app/src/main/java/foundation/e/blisslauncher/features/test/BaseActivity.java
/* Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package foundation.e.blisslauncher.features.test;

import static java.lang.annotation.RetentionPolicy.SOURCE;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.Intent;
import android.view.View.AccessibilityDelegate;
import androidx.annotation.IntDef;
import java.lang.annotation.Retention;
import java.util.ArrayList;

public abstract class BaseActivity extends Activity {

    public static final int INVISIBLE_BY_STATE_HANDLER = 1 << 0;
    public static final int INVISIBLE_BY_APP_TRANSITIONS = 1 << 1;
    public static final int INVISIBLE_ALL =
        INVISIBLE_BY_STATE_HANDLER | INVISIBLE_BY_APP_TRANSITIONS;

    @Retention(SOURCE)
    @IntDef(
        flag = true,
        value = {INVISIBLE_BY_STATE_HANDLER, INVISIBLE_BY_APP_TRANSITIONS})
    public @interface InvisibilityFlags {
    }

    private final ArrayList<VariantDeviceProfile.OnDeviceProfileChangeListener> mDPChangeListeners =
        new ArrayList<>();

    protected VariantDeviceProfile mDeviceProfile;
    protected SystemUiController mSystemUiController;

    private static final int ACTIVITY_STATE_STARTED = 1 << 0;
    private static final int ACTIVITY_STATE_RESUMED = 1 << 1;

    private static final int ACTIVITY_STATE_USER_ACTIVE = 1 << 2;

    @Retention(SOURCE)
    @IntDef(
        flag = true,
        value = {ACTIVITY_STATE_STARTED, ACTIVITY_STATE_RESUMED, ACTIVITY_STATE_USER_ACTIVE})
    public @interface ActivityFlags {
    }

    @ActivityFlags
    private int mActivityFlags;

    // When the recents animation is running, the visibility of the Launcher is managed by the
    // animation
    @InvisibilityFlags
    private int mForceInvisible;

    public VariantDeviceProfile getDeviceProfile() {
        return mDeviceProfile;
    }

    public AccessibilityDelegate getAccessibilityDelegate() {
        return null;
    }

    public static BaseActivity fromContext(Context context) {
        if (context instanceof BaseActivity) {
            return (BaseActivity) context;
        }
        return ((BaseActivity) ((ContextWrapper) context).getBaseContext());
    }

    public SystemUiController getSystemUiController() {
        if (mSystemUiController == null) {
            mSystemUiController = new SystemUiController(getWindow());
        }
        return mSystemUiController;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    protected void onStart() {
        mActivityFlags |= ACTIVITY_STATE_STARTED;
        super.onStart();
    }

    @Override
    protected void onResume() {
        mActivityFlags |= ACTIVITY_STATE_RESUMED | ACTIVITY_STATE_USER_ACTIVE;
        super.onResume();
    }

    @Override
    protected void onUserLeaveHint() {
        mActivityFlags &= ~ACTIVITY_STATE_USER_ACTIVE;
        super.onUserLeaveHint();
    }

    @Override
    protected void onStop() {
        mActivityFlags &= ~ACTIVITY_STATE_STARTED & ~ACTIVITY_STATE_USER_ACTIVE;
        mForceInvisible = 0;
        super.onStop();
    }

    @Override
    protected void onPause() {
        mActivityFlags &= ~ACTIVITY_STATE_RESUMED;
        super.onPause();

        // Reset the overridden sysui flags used for the task-swipe launch animation, we do this
        // here instead of at the end of the animation because the start of the new activity does
        // not happen immediately, which would cause us to reset to launcher's sysui flags and then
        // back to the new app (causing a flash)
        getSystemUiController().updateUiState(SystemUiController.UI_STATE_NORMAL, 0);
    }

    public boolean isStarted() {
        return (mActivityFlags & ACTIVITY_STATE_STARTED) != 0;
    }

    public boolean hasBeenResumed() {
        return (mActivityFlags & ACTIVITY_STATE_RESUMED) != 0;
    }

    public boolean isUserActive() {
        return (mActivityFlags & ACTIVITY_STATE_USER_ACTIVE) != 0;
    }

    public void addOnDeviceProfileChangeListener(VariantDeviceProfile.OnDeviceProfileChangeListener listener) {
        mDPChangeListeners.add(listener);
    }

    public void removeOnDeviceProfileChangeListener(VariantDeviceProfile.OnDeviceProfileChangeListener listener) {
        mDPChangeListeners.remove(listener);
    }

    protected void dispatchDeviceProfileChanged() {
        for (int i = mDPChangeListeners.size() - 1; i >= 0; i--) {
            mDPChangeListeners.get(i).onDeviceProfileChanged(mDeviceProfile);
        }
    }

    public void addForceInvisibleFlag(@InvisibilityFlags int flag) {
        mForceInvisible |= flag;
    }

    public void clearForceInvisibleFlag(@InvisibilityFlags int flag) {
        mForceInvisible &= ~flag;
    }

    public boolean isForceInvisible() {
        return mForceInvisible != 0;
    }
}

File: blisslauncherv2/src/main/java/foundation/e/blisslauncher/base/BaseActivity.kt
package foundation.e.blisslauncher.base

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import androidx.annotation.IntDef
import foundation.e.blisslauncher.common.util.SystemUiController
import javax.inject.Inject

open class BaseActivity : Activity() {

    /*val dpChangeListeners = ArrayList<DeviceProfile.OnDeviceProfileChangeListener>()

    @Inject
    lateinit var deviceProfile: DeviceProfile*/

    @Inject
    lateinit var systemUiController: SystemUiController

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(
        flag = true,
        value = [ACTIVITY_STATE_STARTED, ACTIVITY_STATE_RESUMED, ACTIVITY_STATE_USER_ACTIVE]
    )
    annotation class ActivityFlags

    @ActivityFlags
    private var activityFlags: Int = 0

    val isStarted: Boolean
        get() = activityFlags and ACTIVITY_STATE_STARTED != 0

    val hasBeenResumed: Boolean
        get() = activityFlags and ACTIVITY_STATE_RESUMED != 0

    override fun onStart() {
        activityFlags = activityFlags or ACTIVITY_STATE_STARTED
        super.onStart()
    }

    override fun onResume() {
        activityFlags = activityFlags or ACTIVITY_STATE_RESUMED or ACTIVITY_STATE_USER_ACTIVE
        super.onResume()
    }

    override fun onUserLeaveHint() {
        activityFlags = activityFlags and ACTIVITY_STATE_USER_ACTIVE.inv()
        super.onUserLeaveHint()
    }

    override fun onPause() {
        activityFlags = activityFlags and ACTIVITY_STATE_RESUMED.inv()
        super.onPause()
    }

    override fun onStop() {
        super.onStop()
        activityFlags =
            activityFlags and ACTIVITY_STATE_STARTED.inv() and ACTIVITY_STATE_USER_ACTIVE.inv()
    }

    protected fun dispatchDeviceProfileChanged() {
        //dpChangeListeners.forEach { it.onDeviceProfileChanged(deviceProfile) }
    }

    companion object {
        private const val ACTIVITY_STATE_STARTED = 1 shl 0
        private const val ACTIVITY_STATE_RESUMED = 1 shl 1
        private const val ACTIVITY_STATE_USER_ACTIVE = 1 shl 2

        fun fromContext(context: Context): BaseActivity =
            if (context is BaseActivity) context
            else ((context as ContextWrapper).baseContext) as BaseActivity
    }
}